modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTableLFU.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.siddhi.core.table;
+
+import io.siddhi.core.config.SiddhiAppContext;
+import io.siddhi.core.event.ComplexEventChunk;
+import io.siddhi.core.event.state.StateEvent;
+import io.siddhi.core.event.stream.StreamEvent;
+import io.siddhi.core.table.holder.IndexEventHolder;
+import io.siddhi.core.table.holder.IndexedEventHolder;
+import io.siddhi.core.util.collection.AddingStreamEventExtractor;
+import io.siddhi.core.util.collection.operator.CompiledCondition;
+import io.siddhi.core.util.collection.operator.Operator;
+import io.siddhi.query.api.definition.Attribute;
+import io.siddhi.query.api.definition.TableDefinition;
+
+import static io.siddhi.core.util.SiddhiConstants.CACHE_TABLE_COUNT_LFU;
+import static io.siddhi.core.util.SiddhiConstants.CACHE_TABLE_TIMESTAMP_ADDED;
+import static io.siddhi.core.util.cache.CacheUtils.getPrimaryKey;
+import static io.siddhi.core.util.cache.CacheUtils.getPrimaryKeyFromMatchingEvent;
+
+/**
+ * cache table with FIFO entry removal
+ */
+public class CacheTableLFU extends CacheTable {
+
+    @Override
+    public StreamEvent find(CompiledCondition compiledCondition, StateEvent matchingEvent) {
+        TableState state = stateHolder.getState();
+        readWriteLock.readLock().lock();
+        try {
+            StreamEvent foundEvent = ((Operator) compiledCondition).find(matchingEvent, state.getEventHolder(),
+                    tableStreamEventCloner);
+            String primaryKey;
+
+            if (stateHolder.getState().getEventHolder() instanceof IndexEventHolder) {
```

Comment:
This only support IndexEventHolder.
What happens when the primay key is not used. This class is not handling that. 
Can't we use overwrite(StreamEvent streamEvent) in the IndexEventHolder? Then we don't need to worry about the IndexEventHolder internals

Reply:
inside this if-condition I am updating the latest usage of a particular row. The cache policies have an effect only when store size is greater than cache size. In that case only when it is index event holder (i.e when primary keys are used) the queries are routed to cache. Only in those cases we need to update the last usage. I can add another condition to check if the store is greater than cache before updating last usage. Only information we will be missing here is if store was smaller than cache for some period of time and then it grows the previous usage data wont be available. This will again be irrelevant if cache expiry is enabled. So only disadvantage is when cache expiry is disabled and table grows from being smaller than cache to greater.

If we need to capture that information, we need to update last usages on all the hits. I.e when table is smaller than cache all queries are allowed on cache and results can be more than one. We need to find all the results and update. and without primary keys it is difficult to locate. We need to have a chat regarding this

Reply:
Correction: these functions won't be called if the cache table doesn't have primary keys. I.e only when it is index event holder these functions will be called as per our conditions to check on cache